py/src/call.py

```diff
@@ -606,22 +663,68 @@ async def do_call(
     from starkware.starkware_utils.commitment_tree.patricia_tree.patricia_tree import (
         PatriciaTree,
     )
+    from starkware.starkware_utils.commitment_tree.patricia_tree.nodes import (
+        EmptyNodeFact,
+    )
     from starkware.starknet.services.api.contract_class import EntryPointType
     from starkware.cairo.lang.vm.crypto import pedersen_hash_func
     from starkware.starknet.testing.state import create_invoke_function
+    from starkware.starknet.storage.starknet_storage import StorageLeaf
+    from starkware.starknet.business_logic.state.objects import (
+        ContractState,
+        ContractCarriedState,
+    )
 
     # hook up the sqlite adapter
     ffc = FactFetchingContext(storage=adapter, hash_func=pedersen_hash_func)
 
     # the root tree has to always be height=251
-    shared_state = SharedState(PatriciaTree(root=root, height=251), block_info)
+    shared_state = SharedState(PatriciaTree(root, 251), block_info)
+
+    # we firstly want to load the target contract
+    contract_addresses = {contract_address}
+    # union all those contracts which have pending updates for
+    contract_addresses = contract_addresses | pending_updates.keys()
+    # exclude pending deploys, which should cause an issue reading them
+    contract_addresses = contract_addresses ^ pending_deployed.keys()
+
+    class_hashes = set()
+
+    # include called contract's class hash to be loaded, it doesn't seem to be always
+    # automatically loaded, see test `test_call_on_pending_deployed`.
+    if contract_address in pending_deployed:
+        class_hashes.add(pending_deployed[contract_address])
+
+    # FIXME: it's unknown if this is really required, but we might end up loading up a lot of classes for nothing
+    for class_hash in pending_deployed.values():
+        class_hashes.add(class_hash)
```

Comment:
Probably required for those library call's that we had issue with before?

(Haven't yet checked where this gets used, so just assuming)

Reply:
Those should not be an issue, though I don't have a test case, but they'll be automatically loaded in the normal case. With the two test cases it seemed that new deployed contracts don't automatically get their contract defs loaded, at least always. Might be they worked in the goerli test case because the updated contracts already triggered the contract def loading...

Reply:
To be more precise what I observed is:

1. "tail case" of `test_positive_streamed_on_early_goerli_block_with_deployed` did not need loading all deployed classes' contract hashes
2. `test_call_on_pending_deployed` required loading all deployed classes' contract hashes

Thinking this more I think this is me just missing that contract state updates for other contracts had already forced loading of the contract_hash in 1st case, but on minimal 2nd case there are no "other contracts" to suggest loading of class hashes.

This "suggestion" must happen through loading of each updated contract state, which would link to contract definitions as well at the leaf of global state tree. Yeah, I'm pretty sure this is the reason.

Still thinking this should be on-demand and for all, never for some, but we cannot fix that.

Reply:
Tried to include this in a comment at tail case.